fix: Prevent rendering sky when fog distance is reduced

bytzo opened the pull request:
### Overview

When using fog occlusion, it is possible for the sky to be visible through culled chunks, ruining the illusion that there is fog. This is because vanilla relies on chunk fog to hide the sky when fog distance is reduced, however Sodium's chunk occlusion will cull these chunks as they are completely obscured by fog.

![fog_occlusion_issue](https://user-images.githubusercontent.com/69266322/123052828-1aae3600-d3b8-11eb-83df-234545ff3d28.png)

To solve this issue without completely disabling chunk occlusion, the sky will no longer render when the fog distance is reduced from the default, since more dense fog means you cannot see the sky past it. Without the sky being rendered, the background of the world becomes the exact same as the fog color, so chunks culled by fog occlusion will now be indistinguishable from the fog. This behavior deviates from vanilla and can cause visual differences, however it solves some visual inconsistencies and matches Minecraft Bedrock Edition (it also seems to hide the sky when the fog distance is reduced). The most notable visual differences from vanilla include the sky not being visible from underwater and the End dimension's noise sky not being visible during the boss fight, both due to the fog being reduced.

This fixes #35, fixes #438, and fixes #612. It also fixes [MC-152504](https://bugs.mojang.com/browse/MC-152504) and [MC-230905](https://bugs.mojang.com/browse/MC-230905).

### Screenshots

| Sodium | Sodium + this pull request |
| - | - |
| ![flat_blindness_before](https://user-images.githubusercontent.com/69266322/123053255-8e504300-d3b8-11eb-8d4d-212a1da0db42.png) | ![flat_blindness_after](https://user-images.githubusercontent.com/69266322/123053277-95775100-d3b8-11eb-9a75-51434003c14a.png) |
| ![underwater_before](https://user-images.githubusercontent.com/69266322/123053744-0f0f3f00-d3b9-11eb-8ef9-16be08de6136.png) | ![underwater_after](https://user-images.githubusercontent.com/69266322/123053760-15052000-d3b9-11eb-8d6e-c388186829a5.png) |
| ![end_before](https://user-images.githubusercontent.com/69266322/123053349-a58f3080-d3b8-11eb-8284-012be2dbc362.png) | ![end_after](https://user-images.githubusercontent.com/69266322/123053371-acb63e80-d3b8-11eb-94a8-402c77dfa0e9.png) |
| ![ocean_before](https://user-images.githubusercontent.com/69266322/123053450-c061a500-d3b8-11eb-943c-88a9d2e0dcbc.png) | ![ocean_after](https://user-images.githubusercontent.com/69266322/123053472-c6f01c80-d3b8-11eb-973f-0a92f894f6a5.png) |

bytzo committed:
fix: Prevent rendering sky when fog distance is reduced
When using fog occlusion, it is possible for the sky to be visible through culled chunks, ruining the illusion that there is fog. This is because vanilla relies on chunk fog to hide the sky when fog distance is reduced, however Sodium's chunk occlusion will cull these chunks as they are completely obscured by fog.
To solve this issue without completely disabling chunk occlusion, the sky no longer renders when the fog distance is reduced from the default, since more dense fog means you cannot see the sky past it. Without the sky being rendered, the background of the world becomes the exact same as the fog color, so chunks culled by fog occlusion will now be indistinguishable from the fog. This behavior deviates from vanilla and can cause visual differences, however it solves some visual inconsistencies and matches Minecraft Bedrock Edition (it also seems to hide the sky when the fog distance is reduced). The most notible visual differences from vanilla incude the sky not being visible from underwater and the End dimension's noise sky not being visible during the boss fight, both due to the fog being reduced.
This fixes #35, #438, and #612. It also fixes MC-152504.

bytzo committed:
Merge branch '1.16.x/next' into 1.16.x/fog-occlusion-sky-fix

bytzo committed:
fix: Use LegacyFogHelper instead of GlFogHelper

bytzo committed:
Merge remote-tracking branch 'origin/1.17.x/dev' into 1.17.x/fog-occlusion-sky-fix

bytzo committed:
change: Update to 1.17

bytzo committed:
change: Bump Fabric Loader to 0.11.6

bytzo committed:
change: Set source and target compatibility to Java 16

bytzo committed:
fix: Use camera status instead of fog distance to detect reduced fog
Before, the fog distance was used to detect if the fog was reduced. While this seems like a good solution, the fog distance is calculated after the sky is rendered, meaning the visibility of the sky was always 1 frame behind. Instead, we can use the camera to calculate if the fog distance is reduced, so the visibility of the sky is up-to-date.

bytzo committed:
Merge remote-tracking branch 'origin/1.17.x/dev' into 1.17.x/fog-occlusion-sky-fix

MeeniMc commented:
Just tried this, I love it for the massive visual improvement it brings in when under water (w.r.t. normal Sodium)! 

The loss of the sky rendering in the end is a bit more disruptive than from under water. The  'thick-fog'  effect removes the 'random noise' end sky from view completely. It does however solve the jagged transition between rendered / non-rendered chunks that Sodium also causes in this dimension otherwise, just wish one could have pie and eat it too. 

The 'thick-fog' behavior doesn't result in any discernible rendering difference w.r.t. vanilla in the nether.

wb1016 commented:
used about a week. no bugs directly related encountered.
had some issues with a utility mod that features FreeCam(moving view point without moving actual player position), but it is because they use non-standard rendering. 
according to log, when i enter to liquid in freecam mode, the utility mod write error messeage on the log and halts game.
not sure if it only this branch affects but still it is their issue, not sodium.

jellysquid3 committed:
Merge branch '1.17.x/dev' into 1.17.x/fog-occlusion-sky-fix

bytzo committed:
Rewrite portion of comment describing the patch

I accidentally wrote the exact opposite of the intended message. This commit fixes that, and makes the comment easier to understand by removing unneeded information.
